http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java

```diff
@@ -117,19 +118,29 @@ private NettyNioAsyncHttpClient(DefaultBuilder builder, AttributeMap serviceDefa
     @SdkTestInternalApi
     NettyNioAsyncHttpClient(SdkEventLoopGroup sdkEventLoopGroup,
                             SdkChannelPoolMap<URI, ? extends SdkChannelPool> pools,
-                            NettyConfiguration configuration) {
+                            NettyConfiguration configuration,
+                            ProtocolNegotiation protocolNegotiation) {
         this.sdkEventLoopGroup = sdkEventLoopGroup;
         this.pools = pools;
         this.configuration = configuration;
+        this.protocolNegotiation = protocolNegotiation;
     }
 
     @Override
     public CompletableFuture<Void> execute(AsyncExecuteRequest request) {
+        failIfAlpnUsedWithHttp(request);
         RequestContext ctx = createRequestContext(request);
         ctx.metricCollector().reportMetric(HTTP_CLIENT_NAME, clientName()); // TODO: Can't this be done in core?
         return new NettyRequestExecutor(ctx).execute();
     }
 
+    private void failIfAlpnUsedWithHttp(AsyncExecuteRequest request) {
+        if (protocolNegotiation == ProtocolNegotiation.ALPN
+            && "http".equals(request.request().protocol())) {
+            throw new UnsupportedOperationException("ALPN can only be used with HTTPS, not HTTP.");
```

Comment:
Should we complete the future exceptionally instead? What happens if we throw exception here? Does SDK core handle it correctly?

Reply:
Yes, SDK core handles throwing exception correctly and fails right away. Returned a future completed exceptionally will cause the request to hang